emloop/tests/cli/common_test.py

```diff
@@ -239,3 +278,23 @@ def test_create_model(tmpdir):
     restored_model = create_model(config=new_config, output_dir=tmpdir + '_restored', dataset=dataset,
                                   restore_from=tmpdir)
     assert isinstance(restored_model, DummyModelWithKwargs2)
+
+
+def test_config_file_is_unchanged(tmpdir):
+    """Test that config file is not changed during training."""
+
+    config = {'dataset': {'class': 'emloop.tests.cli.common_test.DummyConfigDataset', 'batch_size': 10,
+                                   'dataset_config': ['a', 'b', 'c']},
+              'stream': {'train': {'rotate': 20}},
+              'hooks': [{'emloop.tests.cli.common_test.DummyConfigHook': {'additional_arg': 10,
+                                                                          'variables': ['a', 'b']}},
+                        {'StopAfter': {'epochs': 1}}],
+              'model': {'class': 'emloop.tests.cli.common_test.DummyConfigModel',
+                                 'architecture': {'model_config': ['a', 'b', 'c', 'd']},
+                                 'io': {'in': [], 'out': ['dummy']}}}
+
+    run(config=config, output_root=tmpdir)
+
+    assert config['dataset']['dataset_config'] == ['a', 'b', 'c']
+    assert config['hooks'][0]['emloop.tests.cli.common_test.DummyConfigHook']['variables'] == ['a', 'b']
+    assert config['model']['architecture']['model_config'] == ['a', 'b', 'c', 'd']
```

Comment:
You can just pass deepcopy() to the run() function and test that the entire dict has not changed.